roadmap/implementers-guide/src/types/candidate.md

```diff
@@ -159,6 +167,52 @@ struct TransientValidationData {
 	///
 	/// This informs a relay-chain backing check and the parachain logic.
 	code_upgrade_allowed: Option<BlockNumber>,
+	/// A copy of `config.max_upward_message_num_per_candidate` for checking that a candidate doesn't
+	/// send more messages that permitted.
+	config_max_upward_message_num_per_candidate: u32,
+	/// The number of messages pending of the downward message queue.
+	dmq_length: u32,
+	/// A part of transient validaiton data related to HRMP.
+	hrmp: HrmpTransientValidationData,
+}
+
+struct HrmpTransientValidationData {
+	/// A vector that enumerates the list of blocks in which there was at least one HRMP message
+	/// received.
+	///
+	/// The first number in the vector, if any, is always greater than the HRMP watermark. The
+	/// elements are ordered by ascending the block number. The vector doesn't contain duplicates.
+	digest: Vec<BlockNumber>,
+	/// The watermark of the HRMP. That is, the block number up to which (inclusive) all HRMP messages
+	/// sent to the parachain are processed.
+	watermark: BlockNumber,
+	/// A mapping that specifies if the parachain can send an HRMP message to the given recipient
+	/// channel. A candidate can send a message only to the recipients that are present in this
+	/// mapping. The number elements in this vector corresponds to the number of egress channels.
+	/// Since it's a mapping there can't be two items with same `ParaId`.
+	egress_limits: Vec<(ParaId, HrmpChannelLimits)>,
+	/// A vector of paras that have a channel to this para. The number of elements in this vector
+	/// correponds to the number of egress channels.
+	ingress_senders: Vec<ParaId>,
+	/// A vector of open requests in which the para participates either as sender or recipient. The
+	/// items are ordered ascending by `HrmpChannelId`. The vector doesn't contain two entries
+	/// with the same `HrmpChannelId`.
+	open_requests: Vec<(HrmpChannelId, HrmpAbridgedOpenChannelRequest)>,
```

Comment:
I wonder if this and close requests need to be part of `PersistedValidationData`. We don't want the collator to be able to forge this data.

Or maybe the hash of that data should be in the `PersistedValidationData`.

Reply:
Right, the PVF needs a way to authenticate that the following events:

1. there is a new inbound open request
1. an outbound open request was confirmed by the recipient
1. an outbound open request timed out. This one could theoretically be tracked by the PVF, but only assuming that the configuration doesn't change which can happen even if rarely.

(I believe other cases can be tracked by the PVF itself. E.g. it should know that it shouldn't send two open requests to the same recipient, confirm the same inbound open request twice, etc)

A collator should not be able to conceal any of these events.

I see the following solutions:

1. put raw requests into `PersistedValidationData`. That inflates the persistent data, but not that bad: the number of open and close requests are bounded (the former is by the sum of max ingress and egress channels, the latter by the number of already opened channels). 
    - Then we can further compress data by leveraging the fact that either sender or recipient is this para. 
    - Then, we can remove `confirmed: bool` for all inbound open requests.
    - This data will be duplicated for successive block in the same session though.
1. put a hash into `PersistedValidationData`. 
    - Essentially, allows us to move this data into the PVF's state and not duplicate it.
1. send a DM for each of these events.
    - The same as previous but less complex since it relies on an existing mechanism
    - A caveat: the para can be a bit slow to process DMQ and potentially miss a notification or receive it when it's too late.
1. punt off on the relay chain storage root proofs.

Reply:
If we assume that with #1642 we limit the total number of channels to a 2 digit number and then retire HRMP completely may be it makes sense to actually go with 1 for the time being.

Reply:
I created #1663 to address this